bika/lims/browser/worksheet/views/analyses.py

```diff
@@ -71,216 +71,372 @@ def __init__(self, context, request):
-            x = pos_items[0]
-            if pos in empties:
+        layout = layout and layout or []
+        # Map the analysis uids with their positions.
+        occupied = []
+        next_positions = {}
+        for item in layout:
+            uid = item.get('analysis_uid', '')
+            slot = int(item['position'])
+            occupied.append(slot)
+            position = next_positions.get(slot, 1)
+            str_position = "{:010}:{:010}".format(slot, position)
+            next_positions[slot] = position + 1
+            uids_positions[uid] = str_position
+
+        # Fill empties
+        last_slot = max(occupied)
+        empties = [num for num in range(1, last_slot) if num not in occupied]
+        for empty_slot in empties:
+            str_position = "{:010}:{:010}".format(empty_slot, 1)
+            uid = "empty-{}".format(empty_slot)
+            uids_positions[uid] = str_position
+
+        return uids_positions
+
+    def get_item_position(self, analysis_uid):
+        """
+        Returns a list with the position for the analysis_uid passed in within
+        the current worksheet in accordance with the current layout, where
+        the first item from the list returned is the slot and the second is
+        the position of the analysis within the slot.
+        :param analysis_uid: uid of the analysis the position is requested
+        :return: the position (slot + position within slot) of the analysis
+        :rtype: list
+        """
+        str_position = self.uids_strpositions.get(analysis_uid, '')
+        tokens = str_position.split(':')
+        if len(tokens) != 2:
+            return None
+        return [to_int(tokens[0]), to_int(tokens[1])]
+
+    def get_item_slot(self, analysis_uid):
+        """
+        Returns the slot number where the analysis must be rendered.
+        An slot contains all analyses that belong to the same parent
+        (AnalysisRequest, ReferenceSample).
+        :param analysis_uid: the uid of the analysis the slot is requested
+        :return: the slot number where the analysis must be rendered
+        """
+        position = self.get_item_position(analysis_uid)
+        if not position:
+            return None
+        return position[0]
+
+    def _get_slots(self, empty_uid=False):
+        """
+        Returns a list with the position number of the slots that are occupied
+        (if empty_uid=False) or are empty (if empty_uid=True)
+        :param empty_uid: True exclude occupied slots. False excludes empties
+        :return: sorted list with slot numbers
+        """
+        slots = list()
+        for uid, position in self.uids_strpositions.items():
+            if empty_uid and not uid.startswith('empty-'):
+                continue
+            elif not empty_uid and uid.startswith('empty-'):
+                continue
+            tokens = position.split(':')
+            slots.append(to_int(tokens[0]))
+        return list(set(slots))
```

Comment:
Not sure if the order of the slots matter, because afaik a `set` is unordered.

Reply:
```
>>> test1 = [5,6,4,7,3,2,1]
>>> test2 = [1,2,3,4,5,6,7]
>>> test3 = [7,6,5,4,3,2,1]
>>> test4 = [5,5,6,6,4,4,7,7,3,3,2,2,1,1]
>>> test5 = [5,4,2,5,4,3,1,2,6,3,7,6,7,1]
>>> list(set(test1))
[1, 2, 3, 4, 5, 6, 7]
>>> list(set(test2))
[1, 2, 3, 4, 5, 6, 7]
>>> list(set(test3))
[1, 2, 3, 4, 5, 6, 7]
>>> list(set(test4))
[1, 2, 3, 4, 5, 6, 7]
>>> list(set(test5))
[1, 2, 3, 4, 5, 6, 7]
```

Reply:
Cool, thanks for that test, I wasn't sure about the actual behavior. This means now a `set` does not maintain the initial order, but returns it sorted. Just also checked this regarding alphabetical vs. literal sort:

```
In [52]: set([-1, 1, 2, 3, 11, 12, 13])
Out[52]: {-1, 1, 2, 3, 11, 12, 13}
```